cli/command/image/build.go

```diff
@@ -235,6 +247,47 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		}
 	}
 
+	// replace Dockerfile if added dynamically
+	if dockerfileCtx != nil {
+		file, err := ioutil.ReadAll(dockerfileCtx)
+		dockerfileCtx.Close()
+		if err != nil {
+			return err
+		}
+		now := time.Now()
+		hdrTmpl := &tar.Header{
+			Mode:       0600,
+			Uid:        0,
+			Gid:        0,
+			ModTime:    now,
+			Typeflag:   tar.TypeReg,
+			AccessTime: now,
+			ChangeTime: now,
+		}
+		randomName := ".dockerfile." + stringid.GenerateRandomID()[:12]
+
+		buildCtx = archive.ReplaceFileTarWrapper(buildCtx, map[string]archive.TarModifierFunc{
+			randomName: func(_ string, h *tar.Header, content io.Reader) (*tar.Header, []byte, error) {
+				return hdrTmpl, file, nil
+			},
+			".dockerignore": func(_ string, h *tar.Header, content io.Reader) (*tar.Header, []byte, error) {
+				if h == nil {
+					h = hdrTmpl
+				}
+				b := &bytes.Buffer{}
+				if content != nil {
+					_, err := b.ReadFrom(content)
+					if err != nil {
+						return nil, nil, err
+					}
+				}
+				b.Write([]byte("\n" + randomName))
```

Comment:
first a nit... can you add `\n` to the end of this line so that if someone does a `cat .dockerignore` it doesn't look like its missing a CR at the end.

More importantly, iff .dockerignore did not already exist then I think we need to also add `.dockerignore` to `.dockerignore`.  This is so that we're not leaving around extra files in the user's context that they don't want - even hidden files.  Granted we should also try to remove these extra lines from the file that we're adding but that's a lot harder and at this time probably not worth the effort - until someone complains. 

Also, in my testing I'm seeing the random file appear but only when I already have a `.dockerignore` with `.dockerignore` in it:
```
$ docker build -f - . < ddd
Sending build context to Docker daemon  6.144kB
Step 1/3 : FROM ubuntu
 ---> 0ef2e08ed3fa
Step 2/3 : COPY * /tmp/
 ---> 93546ad88407
Removing intermediate container 9a87cad3d2bd
Step 3/3 : RUN ls -alt /tmp/
 ---> Running in 278f61144204
total 20
drwxr-xr-x 35 root root 4096 Mar 10 13:34 ..
drwxrwxrwt  2 root root 4096 Mar 10 13:34 .
-rw-------  1 root root   46 Mar 10 13:34 .dockerfile.3146c886e77c
-rw-r--r--  1 root root   46 Mar 10 13:34 ddd
-rw-r--r--  1 root root   19 Mar 10 13:25 Dockerfile
 ---> 2b8485d37faf
Removing intermediate container 278f61144204
Successfully built 2b8485d37faf

$ cat ddd
FROM ubuntu
COPY * /tmp/
RUN ls -alt /tmp/
```

Interestingly, when I add Dockerfile to .dockerignore its not removed either - and this does appear to work on master so something in this PR changed the .dockerignore processing logic.